elements/howto-menu/howto-menu.js

```diff
@@ -0,0 +1,213 @@
+  /**
+   * Define key codes to help with handling keyboard events.
+   */
+  const KEYCODE = {
+    DOWN: 40,
+    ESCAPE: 27,
+    TAB: 9,
+    UP: 38,
+  };
+
+  class HowtoMenu extends HTMLElement {
+    /**
+     * A getter for the first child which is a menuitem.
+     */
+    get _firstMenuItem() {
+      let child = this.firstElementChild;
+      while (child) {
+        if (this._isMenuItem(child)) {
+          return child;
+        }
+        child = next.nextElementSibling;
+      }
+      return null;
+    }
+
+    /**
+     * A getter for the last child which is a menuitem.
+     */
+    get _lastMenuItem() {
+      let child = this.lastElementChild;
+      while (child) {
+        if (this._isMenuItem(child)) {
+          return child;
+        }
+        child = next.previousElementSibling;
+      }
+      return null;
+    }
+
+    /**
+     * Checks if a node is a 'menuitem', 'menuitemcheckbox' or 'menuitemradio'.
+     */
+    _isMenuItem(node) {
+      let ariaRoles = ['menuitem', 'menuitemcheckbox', 'menuitemradio'];
+      return ariaRoles.indexOf(node.getAttribute('role') > -1);
+    }
+
+    /**
+     * Returns a menu item coming after the one passed as an argument, or null.
+     */
+    _nextMenuItem(node) {
+      let next = node.nextElementSibling;
+      while (next) {
+        if (this._isMenuItem(node)) {
+          return next;
+        }
+        next = next.nextElementSibling;
+      }
+      return null;
+    }
+
+    /**
+     * Returns a menu item coming before the one passed as an argument, or null.
+     */
+    _previousMenuItem(node) {
+      let prev = node.previousElementSibling;
+      while (prev) {
+        if (this._isMenuItem(node)) {
+          return prev;
+        }
+        prev = prev.previousElementSibling;
+      }
+      return null;
+    }
+
+    /**
+     * Handles keyboard interaction.
+     */
+    _onKeyDown(event) {
+      let el = event.target;
+      switch (event.keyCode) {
+        case KEYCODE.DOWN:
+          // If arrow down, move to next item. Wrap if necessary.
+          event.preventDefault();
+          let next = this._nextMenuItem(el) || this._firstMenuItem;
+          next.focus();
+          break;
+       case KEYCODE.UP:
+         // If arrow up, move to previous item. Wrap if necessary.
+         event.preventDefault();
+         let prev = this._previousMenuItem(el) || this._lastMenuItem;
+         prev.focus();
+         break;
+       case KEYCODE.ESCAPE:
+         // If escape, close the menu.
+         event.preventDefault();
+         this.opened = false;
+         break;
+       case KEYCODE.TAB:
+         // If escape, close the menu.
+         event.preventDefault();
+         this.opened = false;
+         break;
+       default:
+         break;
+      }
+      // If letter key, move to an item which starts with that letter.
+      if (event.keyCode > 64 && event.keyCode < 91) {
+        for (let i = 0, child; child = this.children[i]; i++) {
```

Comment:
Hm, elegant, but less readable imo (which is arguable, depends on what you're used to I guess). Also would need to convert to array, so the gain is not worth it because perf.